File: todo-reservation-springboot/src/main/java/de/vkb/dojo/es/reservations/rest/ReservationController.java
package de.vkb.dojo.es.reservations.rest;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonProperty;
import de.vkb.dojo.es.common.model.ref.FeedbackReference;
import de.vkb.dojo.es.common.model.ref.Reference;
import de.vkb.dojo.es.reservations.kafka.ReservationCommandProducer;
import de.vkb.dojo.es.reservations.kafka.view.ReservationStore;
import de.vkb.dojo.es.reservations.model.command.CreateReservation;
import de.vkb.dojo.es.reservations.model.command.DeleteReservation;
import de.vkb.dojo.es.reservations.model.command.ReservationCommand;
import de.vkb.dojo.es.reservations.model.dto.ReservationOutput;
import de.vkb.dojo.es.reservations.services.ReferenceService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/reservations")
public class ReservationController {
    final ReferenceService referenceService;
    final ReservationStore reservationStore;
    final ReservationCommandProducer roomCommandProducer;

    public ReservationController(ReferenceService referenceService, ReservationStore reservationStore, ReservationCommandProducer roomCommandProducer) {
        this.referenceService = referenceService;
        this.reservationStore = reservationStore;
        this.roomCommandProducer = roomCommandProducer;
    }

    private ResponseEntity<Reference> sendCommand(CommandOp lambda) {
        var feedback = new FeedbackReference(UUID.randomUUID().toString());
        var command = lambda.op(feedback.getId());
        roomCommandProducer.publish(command);
        return referenceService.sendResponse(HttpStatus.ACCEPTED, feedback);
    }

    @GetMapping(value = "", produces = "application/json")
    public List<ReservationOutput> list() {
        return reservationStore.getAll();
    }

    @PostMapping(value = "", produces = "application/json")
    public ResponseEntity<Reference> create(@RequestBody ReservationCreateData data) {
        return sendCommand( id -> new CreateReservation(
                id,
                data.getRoom(),
                data.getPerson()
        ));
    }

    @GetMapping(value = "/{aggregateId}", produces = "application/json")
    public ReservationOutput read(@PathVariable String aggregateId) {
        return reservationStore.get(aggregateId).orElse(null);
    }

    @DeleteMapping(value = "/{aggregateId}", produces = "application/json")
    public ResponseEntity<Reference> delete(@PathVariable String aggregateId) {
        return sendCommand( operationId -> new DeleteReservation(
                operationId,
                aggregateId
        ));
    }

    static class ReservationCreateData {
        private final String room;
        private final String person;

        @JsonCreator
        public ReservationCreateData(
                @JsonProperty("room") String room,
                @JsonProperty("person") String person
        ) {
            this.room = room;
            this.person = person;
        }

        @JsonGetter
        public String getRoom() {
            return room;
        }

        @JsonGetter
        public String getPerson() {
            return person;
        }
    }

    interface CommandOp {
        ReservationCommand op(String operationId);
    }
}

File: todo-reservation-micronaut/src/main/kotlin/de/vkb/dojo/es/reservations/rest/ReservationController.kt
package de.vkb.dojo.es.reservations.rest

import de.vkb.dojo.es.common.model.ref.FeedbackReference
import de.vkb.dojo.es.common.model.ref.Reference
import de.vkb.dojo.es.reservations.kafka.view.ReservationStore
import de.vkb.dojo.es.reservations.model.command.CreateReservation
import de.vkb.dojo.es.reservations.model.command.DeleteReservation
import de.vkb.dojo.es.reservations.model.command.ReservationCommand
import de.vkb.dojo.es.reservations.model.dto.ReservationOutput
import io.micronaut.http.HttpRequest
import io.micronaut.http.HttpResponse
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.*
import io.micronaut.http.server.util.HttpHostResolver
import io.micronaut.http.uri.UriBuilder
import java.net.URI
import java.util.*

@Controller("/reservations")
class ReservationController(
    private val reservationStore: ReservationStore,
    private val reservationCommandProducer: de.vkb.dojo.es.reservations.kafka.ReservationCommandProducer,
    private val httpHostResolver: HttpHostResolver
) {
    private fun sendCommand(request: HttpRequest<*>, lambda: (id: String) -> ReservationCommand): HttpResponse<Reference> {
        val feedback = FeedbackReference(id = UUID.randomUUID().toString())
        val command = lambda(feedback.id)
        reservationCommandProducer.sendCommand(command.operationId, command)
        return HttpResponse.accepted<Reference?>()
            .header("Location", UriBuilder.of(URI(httpHostResolver.resolve(request))).path(feedback.path).build().toString())
            .body(feedback)
    }

    @Get("/")
    fun list(): List<ReservationOutput> {
        return reservationStore.all()
    }

    @Post("/")
    @Consumes(MediaType.APPLICATION_JSON)
    fun create(request: HttpRequest<*>, @Body data: ReservationCreateData): HttpResponse<Reference> {
        return sendCommand(request) { operationId ->
            CreateReservation(
                operationId = operationId,
                room = data.room,
                person = data.person
            )
        }
    }

    @Get("/{aggregateId}")
    fun read(@PathVariable aggregateId: String): Optional<ReservationOutput> {
        return reservationStore.get(aggregateId)
    }

    @Delete("/{aggregateId}")
    fun delete(request: HttpRequest<*>, @PathVariable aggregateId: String): HttpResponse<Reference> {
        return sendCommand(request) { operationId ->
            DeleteReservation(
                operationId = operationId,
                aggregateId = aggregateId
            )
        }
    }

    data class ReservationCreateData(
        val room: String,
        val person: String
    )
}
